### PR TITLE
Check and update data before explanation

### PR DESCRIPTION
**Pull request by martinju**
Updated!

The point of this PR is to add proper checks that the data being passed to shapr (training data) and explain (testing data) have the correct format, meaning that they contain the right features, have the right class and if they are factor features have the right levels. While this was not crucial when we only supported numerical features, it has become important now that we support factors as well through the ctree approach. If there are differences between what is defined in model object and in the data, the data are either adjusted/updated and a message is provided, or an error is thrown. 

All of this is mainly done through a the function `preprocess_data`: The arguments to that function is the data to check and a list describing the "correct" feature details which is extracted from the model object using a new function `get_model_specs` which is an updated version of the previous `features` function (now obsolete, so deleted). `preprocess_data` calls `get_data_specs` which is a new function that extracts the feature info from the data. The new function `check_features` is the main contribution, and that compares the feature list from the model with that from the data. That function is motivated by the earlier `make_dummies` and `apply_dummies` functions. The function does no re-shuffling, deletion of unused columns or so, but returns a feature_list that is passed on to `update_data` which handles that. The feature list is written to the explainer at the end of `shapr` and passed on to `explain` where the checking and updating is done for the test data.

A bunch of remarks

1. `get_model_specs` calls the function get_supported_models which gives a table with all function classes defined for the functions `model_type`, `predict_model` (and `get_model_specs`) which are needed to run explain a model class. This seems like a valid approach, but I currently have some issues testing custom models as testthat does not see functions that are defined within a testthat environment, and I am probably not allowed to write to .globalenv within the tests. Testing of `get_supported_models` works and is included in the standard test environment, while tests for custom models are for now done manually by running `expect_silent(source("tests/testthat/manual_test_scripts/test_custom_models.R"))` in the console. This works, but at some point we should find a better way to do this.
 
2. The `make_dummies` function now also makes use of the `check_features` function to reuse more code.

3. A full round of linting and styling of the full package will be done in a separate PR after this one, so just ignore all of that for now.

4. There are tests are not great, but covr::covr() shows that the majority of the code is tested, so I think it is fine for now. Ref 
#249 , we should restucture the tests later, but not for this CRAN release.

5. Tests still failing on macOS. Don't understand why. Keep on ignoring for now.

### CODE REVIEW

File: R/explanation.R [hunk 1]
```diff
@@ -16,11 +16,11 @@
...
 #' @param ... Additional arguments passed to \code{\link{prepare_data}}
 #'
-#' @details The most important thing to notice is that \code{shapr} has implemented three different
+#' @details The most important thing to notice is that \code{shapr} has implemented four different
 #' approaches for estimating the conditional distributions of the data, namely \code{"empirical"},
-#' \code{"gaussian"} and \code{"copula"}.
+#' \code{"gaussian"}, \code{"copula"} and \code{"ctree"}.
 #'
-#' In addition to this the user will also have the option of combining the three approaches.
+#' In addition to this the user will also have the option of combining the four approaches.
```
Comment (aredelmeier) [inline R/explanation.R:23]: In addition, the user also has the option of combining the four approaches.
```suggestion
#' In addition to this the user will also have the option of combining the four approaches.
```

Comment (aredelmeier) [inline R/explanation.R:23]:
```suggestion
#' In addition, the user also has the option of combining the four approaches.
```

**Review by aredelmeier — COMMENTED** [review]
I tried to make it through all the code but I did not finish test-explanation.R and test-models.R. I have a tiny number of comments and spelling this to suggest.

File: tests/testthat/test-sampling.R [hunk 2]
```diff
@@ -55,219 +53,227 @@ test_that("Test sample_combinations", {
...
-                    x_train = x_train,
-                    p = length(x_test), sample = TRUE)
+    tree <- list(tree = datact, given_ind = given_ind, dependent_ind = dependent_ind)
 
-  # Test output format ------------------
-  expect_true(data.table::is.data.table(r))
-  expect_equal(ncol(r), m)
-  expect_equal(nrow(r), n_samples)
-  expect_equal(colnames(r), colnames(x_test_dt))
+    # new
```
Comment (aredelmeier) [inline tests/testthat/test-sampling.R:256]:
```suggestion
```

**Review by aredelmeier — APPROVED** [review]
A few more comments. Nothing serious. Mostly spelling/grammar, some notes about ordering tests, some notes about being consistent with commenting, some notes about naming. I did my best to be as thorough as possible but it is possible I missed some things.

File: inst/scripts/example_custom_model.R [hunk 3]
```diff
@@ -48,19 +49,41 @@ predict_model.gbm <- function(x, newdata) {
...
 explanation <- explain(x_test, explainer, approach = "empirical", prediction_zero = p0)
+# Plot results
+plot(explanation)
 
-# Printing the Shapley values for the test data
-explanation$dt
+#### Minimal version of the three required model functions        ####
+#### Note: Working only for this exact version of the model class ####
+#### Avoiding to define get_model_specs skips all feature         ####
+#### consistency checking between your data and model             ####
```
[the remark this answers is not in the document]
Comment (martinju) [reply inst/scripts/example_custom_model.R:75]: Reject. This is the rest of the sentence...

File: R/features.R [hunk 4]
```diff
@@ -215,171 +210,115 @@ helper_feature <- function(m, feature_sample) {
...
+  feature_list <- updater
 
-  # re-level traindata and testdata
-  for (i in names(list_levels_train)) {
-    traindata[[i]] <- factor(traindata[[i]], levels = list_levels_train[[i]])
-  }
-  for (i in names(list_levels_test)) {
-    testdata[[i]] <- factor(testdata[[i]], levels = list_levels_test[[i]])
-  }
+  # Extracts the components that the were used before. Consider cleaning up this.
```
[the remark this answers is not in the document]
Comment (martinju) [reply R/features.R:241]: Just an old comment left behind

File: R/preprocess_data.R [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+    stop(paste0(name_1," must have column names."))
+  }
+  if (is.null(f_list_2$labels)) {
+    stop(paste0(name_2," must have column names."))
+  }
+
+  NA_1 <- any(is.na(f_list_1$labels))
+  NA_2 <- any(is.na(f_list_2$labels))
+
+  if((NA_1 & NA_2) | ((NA_1 | NA_2) & !use_1_as_truth)){
```
[the remark this answers is not in the document]
Comment (martinju) [reply R/preprocess_data.R:159]: The second holds if just one of the f_lists have NA's, while you also have specified use_1_as_truth. Thus, if you are using make_dummies with a proper training set, but where the test set does not have column names for instance.

Are you following?

File: R/preprocess_data.R [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+    stop(paste0(name_1," must have unique column names."))
+  }
+
+  # feature names must be unique
+  if (any(duplicated(f_list_2$labels))) {
+    stop(paste0(name_2," must have unique column names."))
+  }
+
+
+  missing_1_in_2 <- f_list_1$labels[!(f_list_1$labels %in% f_list_2$labels)]
```
[the remark this answers is not in the document]
Comment (martinju) [reply R/preprocess_data.R:179]: Good point, will fix.

File: R/preprocess_data.R [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+  order_1 <- match(f_list_1$labels,names(f_list_1$classes))
+  f_list_1$classes <- f_list_1$classes[order_1]
+  f_list_1$factor_levels <- f_list_1$factor_levels[order_1]
+
+  # for f_list_2
+  order_2 <- match(f_list_2$labels,names(f_list_2$classes))
+  f_list_2$classes <- f_list_2$classes[order_2]
+  f_list_2$factor_levels <- f_list_2$factor_levels[order_2]
+
+  #### Reorder f_List_2 to match f_list_1, also removing anything in the former which is not in the latter ####
```
[the remark this answers is not in the document]
Comment (martinju) [reply R/preprocess_data.R:209]: You are absolutely right! 👍 Will fix

File: R/preprocess_data.R [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+update_data = function(data,updater){
+  # Operates on data by reference, so no copying of data here
+
+  new_labels <- updater$labels
+  factor_levels <- updater$factor_levels
+
+  # Reorder and delete unused columns
+  cnms_remove <- setdiff(colnames(data), new_labels)
+  if (length(cnms_remove) > 0) {
+    message(paste0("The columns(s) ",paste0(cnms_remove,collapse=", ")," is not used by the model and thus removed ",
```
[the remark this answers is not in the document]
Comment (martinju) [reply R/preprocess_data.R:300]: I don't see anything strange here. Can you be more specific?

File: inst/scripts/example_custom_model.R [hunk 9]
```diff
@@ -8,37 +8,38 @@ data("Boston", package = "MASS")
...
 
-xy_train <- tail(Boston, -6)
-x_test <- head(Boston,6)
+x_train <- as.matrix(Boston[-1:-6, x_var])
+y_train <- Boston[-1:-6, y_var]
+x_test <- as.matrix(Boston[1:6, x_var])
 
 form = as.formula(paste0(y_var,"~",paste0(x_var,collapse="+")))
 
+library(gbm)
```
[the remark this answers is not in the document]
Comment (martinju) [reply inst/scripts/example_custom_model.R:17]: Yes, just a script, so we don't need it.

File: inst/scripts/example_custom_model.R [hunk 10]
```diff
@@ -48,19 +49,41 @@ predict_model.gbm <- function(x, newdata) {
...
+#### Minimal version of the three required model functions        ####
+#### Note: Working only for this exact version of the model class ####
+#### Avoiding to define get_model_specs skips all feature         ####
+#### consistency checking between your data and model             ####
 
+predict_model.gbm <- function(x, newdata) {
+  predict(x, as.data.frame(newdata),n.trees = x$n.trees)
+}
+
+rm(get_model_specs.gbm)
```
[the remark this answers is not in the document]
Comment (martinju) [reply inst/scripts/example_custom_model.R:81]: The point here is to show two different ways to define the  functions necessary to explain the gbm model. In the minimalist approach, we don't define the get_model_specs function. I will make it clearer.

File: tests/testthat/manual_test_scripts/test_custom_models.R [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+# Doing all testing from shapr
+
+library(testthat)
+library(shapr)
+library(gbm)
+library(MASS)
+
+  # Data -----------
+  data("Boston", package = "MASS")
```
[the remark this answers is not in the document]
Comment (martinju) [reply tests/testthat/manual_test_scripts/test_custom_models.R:14]: Done

File: tests/testthat/manual_test_scripts/test_custom_models.R [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+      "classification",
+      "regression"
+    )
+    if (model_type == "classification") {
+      predict(x, as.data.frame(newdata), type = "response",n.trees = x$n.trees)
+    } else {
+      predict(x, as.data.frame(newdata),n.trees = x$n.trees)
+    }
+  }
+  expect_silent(shapr(train_df_used_numeric ,model_custom)) # Both defined, so pass silently
```
[the remark this answers is not in the document]
Comment (martinju) [reply tests/testthat/manual_test_scripts/test_custom_models.R:65]: Very good point! It does indeed, which is not good. I guess we should check that it produces a numeric vector. I am adding 

tmp <- predict_model(model, head(x_train,2)) # Checking that the prediction function works
  if(!(all(is.numeric(tmp)) & length(tmp)==2)){
stop(...)
}

File: tests/testthat/test-a-shapley.R [hunk 13]
```diff
@@ -1,77 +1,124 @@
...
+  data_error <- train_df[,-3]
+  expect_error(shapr(data_error,model))
+
+  # Duplicated column names
+  data_error <- train_df_used_factor
+  data_error <- cbind(data_error,lstat = 1)
+  expect_error(shapr(data_error,model))
+
+  # Empty column names in data
+  if (requireNamespace("xgboost", quietly = TRUE)) {
```
[the remark this answers is not in the document]
Comment (martinju) [reply tests/testthat/test-a-shapley.R:104]: Yes, this is not checking for empty column names in the data, but rather in the model. I added a lm example that checks it for the data, but I am considering keeping the xgboost example also. Will have to think through whether we might accept missing feature names in the model in the same way as we accept that there are NA's there -- by switching to trusting the data features in that case.

File: vignettes/understanding_shapr.Rmd [hunk 14]
```diff
@@ -558,79 +558,110 @@ plot(explanation_categorical, plot_phi0 = FALSE, index_x_test = c(1, 6))
...
+#### Minimal version of the three required model functions        ####
+#### Note: Working only for this exact version of the model class ####
+#### Avoiding to define get_model_specs skips all feature         ####
+#### consistency checking between your data and model             ####
+
+predict_model.gbm <- function(x, newdata) {
+    predict(x, as.data.frame(newdata),n.trees = x$n.trees)
+}
+
+rm(get_model_specs.gbm)
```
[the remark this answers is not in the document]
Comment (martinju) [reply vignettes/understanding_shapr.Rmd:653]: See above. Will fix

File: R/preprocess_data.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (aredelmeier) [reply R/preprocess_data.R:159]: Sorry I completely missed the distinction between & and | yesterday so it looked like both had NA_1 & NA_2. I do follow.

File: inst/scripts/example_custom_model.R (hunk 10 above)
[the remark this answers is not in the document]
Comment (aredelmeier) [reply inst/scripts/example_custom_model.R]: I guess my question was we had to rm() at the end of the script. Does rm() effect the shapr() and explain() that come further in the script?

File: R/preprocess_data.R (hunk 8 above)
[the remark this answers is not in the document]
Comment (aredelmeier) [reply R/preprocess_data.R:301]: hmm funny I have no idea what I meant here.

[the remark this answers is not in the document]
Comment (aredelmeier) [reply R/preprocess_data.R:301]: Oh I remember: I just thought it was strange that you put "is not used by the model" and "from the data" on two lines (without a \n). Perhaps you had a reason?

File: tests/testthat/test-explanation.R [hunk 15]
```diff
@@ -22,225 +19,234 @@ test_that("Test get_list_approaches", {
...
-  ex_list[[22]] <- explain(x_test, explainer, approach = "ctree", prediction_zero = p0, sample = TRUE,
-                           mincriterion = 0.9, minsplit = 20, minbucket = 25)
+    if (requireNamespace("party", quietly = TRUE)) {
 
-  # Ex 23: Explain combined - ctree and gaussian, sample = FALSE
-  ex_list[[23]] <- explain(x_test, explainer, approach = c("ctree", rep("gaussian", 3)),
-                           prediction_zero = p0, sample = FALSE)
+      # Ex 18: Explain combined II - all empirical
+      approach <- c(rep("empirical", 4))
+      ex_list[[18]] <- explain(x_test, explainer, approach = approach, prediction_zero = p0)
```
[the remark this answers is not in the document]
Comment (martinju) [reply tests/testthat/test-explanation.R:100]: Good catch. Just a mistake. Will fix

File: tests/testthat/test-explanation.R [hunk 16]
```diff
@@ -22,225 +19,234 @@ test_that("Test get_list_approaches", {
...
-      explain(x_test, explainer, approach = "ctree", prediction_zero = p0, sample = FALSE,
-              mc_cores_create_ctree = 1, mc_cores_sample_ctree = multicore)
-    )
-
-    expect_equal(
-      explain_base_nosample,
-      explain(x_test, explainer, approach = "ctree", prediction_zero = p0, sample = FALSE,
-              mc_cores_create_ctree = multicore, mc_cores_sample_ctree = 1)
-    )
+      # Ex 38: Test that ctree with mincriterion equal to same probability four times gives the same as only passing one
```
[the remark this answers is not in the document]
Comment (martinju) [reply tests/testthat/test-explanation.R:180]: Fixing

File: tests/testthat/test-explanation.R [hunk 17]
```diff
@@ -22,225 +19,234 @@ test_that("Test get_list_approaches", {
...
+        explain_base_sample <- explain(x_test, explainer, approach = "ctree", prediction_zero = p0, sample = TRUE)
+
+        # Seed consistent when only paralellizing create_ctree, and not sample_ctree
+        expect_equal(
+          explain_base_sample,
+          explain(x_test, explainer, approach = "ctree", prediction_zero = p0, sample = TRUE,
+                  mc_cores_create_ctree = multicore, mc_cores_sample_ctree = 1)
+        )
+
+        # Seed consistent, when run twice with same seed
```
[the remark this answers is not in the document]
Comment (martinju) [reply tests/testthat/test-explanation.R:232]: Did "Consistent results when ran twice with same seed" instead

File: tests/testthat/test-explanation.R [hunk 18]
```diff
@@ -249,147 +255,221 @@ test_that("Test functions in explanation.R", {
...
+
+    # Test data
+    all_test_data <- list(
+      x_test,
+      x_test_reordered,
+      x_test_full
+    )
+
+    # Linear model
+    ll <- list(
```
[the remark this answers is not in the document]
Comment (martinju) [reply tests/testthat/test-explanation.R:307]: Yes, did it for this file at least.

File: tests/testthat/test-models.R [hunk 19]
```diff
@@ -1,351 +1,494 @@
...
-    )
+  formula_binary_numeric <- as.formula(paste0("y_train_binary ~ ",paste0(x_var_numeric,collapse="+")))
+  formula_binary_factor <- as.formula(paste0("y_train_binary ~ ",paste0(x_var_factor,collapse="+")))
+
+  dummylist <- make_dummies(traindata = x_train[, x_var_factor], testdata = x_train[, x_var_factor])
+
+  # List of models to run silently
+  l_silent <- list(
+    stats::lm(formula_numeric, data = train_df),
+    stats::glm(formula_numeric, data = train_df),
```
[the remark this answers is not in the document]
Comment (martinju) [reply tests/testthat/test-models.R:274]: Now grouping model classes together.

File: tests/testthat/test-models.R [hunk 20]
```diff
@@ -1,351 +1,494 @@
...
+
+    train_df <- cbind(x_train, y_train)
+    x_var_numeric <- c("lstat", "rm", "dis", "indus")
+    formula_numeric <- as.formula(paste0("y_train ~ ",paste0(x_var_numeric,collapse="+")))
+
+    # Backdoors in main func
 
+    # Unsupported model
+    model_unsupported <- gbm::gbm(formula_numeric,distribution = "gaussian",data = train_df)
+    model_native <- stats::lm(formula_numeric, data = train_df)
```
[the remark this answers is not in the document]
Comment (martinju) [reply tests/testthat/test-models.R:483]: Thanks

**Review by martinju — COMMENTED** [review]
Thanks for a through review, this was just what I wanted. I think have been through it all now. Will fix a few things and then push again.
